src/satosa/micro_services/custom_routing.py

```diff
@@ -2,14 +2,117 @@
 from base64 import urlsafe_b64encode
 
 from satosa.context import Context
+from satosa.internal import InternalData
+
 from .base import RequestMicroService
 from ..exception import SATOSAConfigurationError
 from ..exception import SATOSAError
+from ..exception import SATOSAStateError
 
 
 logger = logging.getLogger(__name__)
 
 
+class CustomRoutingError(SATOSAError):
+    """
+    SATOSA exception raised by CustomRouting rules
+    """
+    pass
+
+
+class DecideBackendByTarget(RequestMicroService):
+    """
+    Select which backend should be used based on who is the SAML IDP
+    """
+
+    def __init__(self, config, *args, **kwargs):
+        """
+        Constructor.
+        :param config: microservice configuration loaded from yaml file
+        :type config: Dict[str, Dict[str, str]]
+        """
+        super().__init__(*args, **kwargs)
+        self.target_mapping = config['target_mapping']
+        self.endpoint_paths = config['endpoint_paths']
+        self.default_backend = config['default_backend']
+
+        if not isinstance(self.endpoint_paths, list):
+            raise SATOSAConfigurationError()
+
+    def register_endpoints(self):
+        """
+        URL mapping of additional endpoints this micro service needs to register for callbacks.
+
+        Example of a mapping from the url path '/callback' to the callback() method of a micro service:
+            reg_endp = [
+                ("^/callback1$", self.callback),
+            ]
+
+        :rtype List[Tuple[str, Callable[[satosa.context.Context, Any], satosa.response.Response]]]
+
+        :return: A list with functions and args bound to a specific endpoint url,
+                 [(regexp, Callable[[satosa.context.Context], satosa.response.Response]), ...]
+        """
+
+        # this intercepts disco response
+        return [
+            (path , self.backend_by_entityid)
+            for path in self.endpoint_paths
+        ]
+
+    def _get_backend(self, context:str, entity_id:str) -> str:
+        """
+        returns the Target Backend to use
+        """
+        return (
+            self.target_mapping.get(entity_id) or
+            self.default_backend
+        )
+
+    def process(self, context, data):
+        """
+        Will modify the context.target_backend attribute based on the target entityid.
+        :param context: request context
+        :param data: the internal request
+        """
+        entity_id = context.request.get('entityID')
+        if entity_id:
+            self._rewrite_context(entity_id, context)
+        return super().process(context, data)
+
+    def _rewrite_context(self, entity_id, context) -> None:
+        tr_backend = self._get_backend(context, entity_id)
+        context.internal_data['target_entity_id'] = entity_id
+        context.target_frontend = context.target_frontend or context.state.get('ROUTER')
```

Comment:
why do we need to set `context.target_frontend`?

Reply:
I used it just to have a way to know that's a flow touched by this microservice :)

Reply:
target_frontend should already be set by the routing module. I don't think we should be changing it here.
This seems unrelated to the goal of this micro-service.

Reply:
Found it, it's only required for log message, here:
https://github.com/IdentityPython/SATOSA/pull/220/files/6ccedc469751e2a441e89fec33e1f0c02a839253#diff-643d6b17c79aec1dad0acafabfca8efe6dd3dcca590bcfda363aa44b06079830R90